Source/Graphic/HCContour.c

```diff
@@ -0,0 +1,394 @@
+}
+
+HCInteger HCContourElementHashValue(HCContourElement element) {
+    return
+        HCPointHashValue(element.p) ^
+        HCPointHashValue(element.c0) ^
+        HCPointHashValue(element.c1);
+}
+
+void HCContourElementPrint(HCContourElement element, FILE* stream) {
+    fprintf(stream, "<p:");
+    HCPointPrint(element.p, stream);
+    fprintf(stream, ",c0:");
+    HCPointPrint(element.c0, stream);
+    fprintf(stream, ",c1:");
+    HCPointPrint(element.c1, stream);
+    fprintf(stream, ">");
+}
+
+//----------------------------------------------------------------------------------------------------------------------------------
+// MARK: - Element Conversion
+//----------------------------------------------------------------------------------------------------------------------------------
+HCContourElement HCContourElementAsLinear(HCContourElement element) {
+    if (HCContourElementIsLinear(element)) {
+        return element;
+    }
+    if (HCContourElementIsQuadratic(element)) {
+        return HCContourElementMakeLinear(element.p);
+    }
+    return HCContourElementMakeLinear(element.p);
+}
+
+HCContourElement HCContourElementAsQuadratic(HCContourElement element) {
+    if (HCContourElementIsQuadratic(element)) {
+        return element;
+    }
+    if (HCContourElementIsCubic(element)) {
+        return HCContourElementMakeQuadratic(element.p, HCPointInterpolate(element.c0, element.c1, 0.5));
+    }
+    return HCContourElementMakeQuadratic(element.p, element.p);
+}
+
+HCContourElement HCContourElementAsCubic(HCContourElement element) {
+    if (HCContourElementIsCubic(element)) {
+        return element;
+    }
+    if (HCContourElementIsQuadratic(element)) {
+        return HCContourElementMakeCubic(element.p, element.c0, element.c0);
+    }
+    return HCContourElementMakeCubic(element.p, element.p, element.p);
+}
+
+//----------------------------------------------------------------------------------------------------------------------------------
+// MARK: - Element Evaluation
+//----------------------------------------------------------------------------------------------------------------------------------
+HCPoint HCContourEvaluateElement(HCReal t, HCPoint p0, HCContourElement element, HCReal* dx, HCReal* dy) {
+    if (HCPointIsInvalid(element.c1)) {
+        if (HCPointIsInvalid(element.c0)) {
+            return HCContourEvaluateLinearCurve(t, p0, element.p, dx, dy);
+        }
+        return HCContourEvaluateQuadraticCurve(t, p0, element.p, element.c0, dx, dy);
+    }
+    return HCContourEvaluateCubicCurve(t, p0, element.p, element.c0, element.c1, dx, dy);
+}
+
+HCPoint HCContourEvaluateLinearCurve(HCReal t, HCPoint p0, HCPoint p1, HCReal* dx, HCReal* dy) {
+    HCReal tc = 1.0 - t;
+        
+    HCReal x = p0.x * tc + p1.x * t;
+    HCReal y = p0.y * tc + p1.y * t;
+        
+//    HCReal a = tc;
+//    HCReal b = t;
+//    HCReal x = a * p0.x + b * p1.x;
+//    HCReal y = a * p0.y + b * p1.y;
```

Comment:
Did you want these comments?

Reply:
Yes, they are the intermediate steps to another method of evaluation. I could just look them up in git, I suppose.